### DISCUSSION TITLE
Solara doesn't render in Databricks notebook

### DISCUSSION THREAD

**Discussion by Ljupch0** [widgetti/solara#128] [General]
Hi Maarten,

Thank you for your work!

In my use case I envisioned Solara as a extra nice UI option in shared Databricks Notebooks running on clusters. A way to give less technical users an interface without deploying separate apps, by using the Notebook environment for both authentication and compute.

However it seems that it cannot render:

![image](https://github.com/widgetti/solara/assets/46574875/a87988d3-d2e0-49f3-bddf-438fbbed94e8)

Is there anything in the limitations section of this page that applies to Solara? https://learn.microsoft.com/en-us/azure/databricks/notebooks/ipywidgets 

Unfortunately I cannot provide a reproducible example :(

**Comment by maartenbreddels** [comment]
Hi Ljupcho,

Thank you for reporting this. Do you see an error in the JS console?

Regards,

Maarten

**Comment by Ljupch0** [comment]
I cleared console before running the chunk with solara code.

There is an error `Unchecked runtime.lastError: The message port closed before a response was received.`, but this same error is printed also when running other, working, ipywidgets. The error is stemming from the iframe.html code pasted below, but the error in the code
'iframe injection failed frameId check' is not the one printed.

It seems to load for a second before the `Cannot show widget` text is displayed.

```html
<!DOCTYPE html>
<html>
    <head>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <script>
            if (!String.prototype.endsWith) { // IE11
                String.prototype.endsWith = function(search, this_len) {
                    if (this_len === undefined || this_len > this.length) {
                        this_len = this.length;
                    }
                    return this.substring(this_len - search.length, this_len) === search;
                };
            }
            const hash = JSON.parse(decodeURIComponent(window.location.hash.substr(1)));
            const userContentHost = '.databricksusercontent.com';
            // only allow if sandboxed w/o allow-same-origin OR is on databricksusercontent.com
            if (!document.domain ||
                (document.domain && document.domain.endsWith(userContentHost))) {
                window.addEventListener('message', function (msg) {
                    if (msg.data.frameId === hash.frameId && msg.data.type === 'injectHtml') {
                        // document.open(); // ommitted as it breaks Edge. see PR 30550
                        document.write(msg.data.content);
                        document.close();
                    } else {
                        console.error('iframe injection failed frameId check');
                    }
                });
                window.parent.postMessage({
                    frameId: hash.frameId,
                    type: 'requestInjection',
                }, hash.parentUri);
            }
        </script>
    </head>
    <body></body>
</html>
```

**Comment by maartenbreddels** [comment]
Does a regular ipywidget work, e.g:
```
import ipywidgets as w
display(w.Button(description="Do you see me?"))
```

**Comment by Ljupch0** [comment]
![image](https://github.com/widgetti/solara/assets/46574875/a52d0f58-c704-41b7-a9f7-0b763989b435)

**Comment by maartenbreddels** [comment]
Can you run two more tests:
```python
import ipyvue
ipyvue.Html(tag="h2", children=["test"])
```

and
```python
import ipyvuetify as v
v.Btn(children=["Do you see me?"])
```

**Comment by Ljupch0** [comment]
These work also
![image](https://github.com/widgetti/solara/assets/46574875/7b860cdf-f737-40bc-9781-d626f593e0e9)

**Comment by maartenbreddels** [comment]
Didn't expect that!

What about?
```python
box, rc = solara.render(Page())
box
```

**Comment by Ljupch0** [comment]
Works!
![image](https://github.com/widgetti/solara/assets/46574875/0288ad5a-49e9-4e30-88ed-c31b33f65810)

**Comment by maartenbreddels** [comment]
O, many thanks!
A shorter path might be `Page.widget()`, but we need to somehow fix this in the long run. Would you mind opening an issue, instead of a discussion?

**Comment by Ljupch0** [comment]
So in the meanwhile I can call all pages as `Page.widget()`? I confirm this also works. Opening issue.

**Reply by maartenbreddels** [reply]
You only need to do that for the 'main widget', inside of the component you do not need that.

**Reply by abelusha** [reply]
@maartenbreddels can you help with this in databricks?
```python
import solara
import solara.lab

df = px.data.iris()
columns = list(df.columns)
x_axis = solara.reactive("sepal_length")
y_axis = solara.reactive("sepal_width")
@solara.component
def Page():
    df = px.data.iris()
    fig = px.scatter(df, "sepal_length", "sepal_width")   
    solara.FigurePlotly(fig)
    solara.Select(label="X-axis", value=x_axis, values=columns)
    solara.Select(label="Y-axis", value=y_axis, values=columns)    


@solara.component
def Workaround():
    with solara.components.applayout._AppLayoutEmbed():
        Page()
Workaround.widget()
```
I am getting this error. "Error displaying widget: Cannot read properties of undefined (reading 'buffer')"

**Reply by maartenbreddels** [reply]
Not yet, for this, I need to have a Databricks account. This takes a bit of time because we also need to set up a new AWS account.

Maybe @jasongrout-db / @jasongrout could help debug this in the meantime since I think it is probably an issue with plotly and binary buffers at Databricks and not a solara issue per se.

**Reply by Ljupch0** [reply]
I am getting the same error every time I include plotly code on databricks

**Comment by jasongrout-db** [comment]
@Ljupch0 - I think it's great that you're bringing this up here, but please also put a request in through your Databricks account rep. That works through our ticketing system and bumps the priority for internal people to work on this. (Disclaimer: I work at Databricks, and would love to see Solara working in it)

**Comment by maartenbreddels** [comment]
@jasongrout-db my best guess here is that on databricks, we mime bundle rendering gives higher priority to HTML, than the jupyter widget. The code path that gets triggered:
 * https://github.com/widgetti/reacton/blob/735eaaf14c7d857fc9194f909b98a2df84c840ae/reacton/core.py#L311
 * https://github.com/widgetti/reacton/blob/735eaaf14c7d857fc9194f909b98a2df84c840ae/reacton/core.py#L2175
 
 Which adds an HTML entry in the mime bundle that @Ljupch0 sees

**Comment by ccdavid** [comment]
The workarounds here didn't work for me 100% of times. It does work right after clearing the state in the notebook, but fails if I re-create the `Page` object (e.g. rerun the cell). Another way to make display work again is by displaying two pages in the same cell. The first attempt will fail, but the second one will show and that will fix the display in all other cells in the notebook.

I did the following workaround to make sure the page is always displayed regardless of the state in the notebook:

```
def workarond_display(solara_component):
  # display component in an output that won't be displayed to notebook output
  # and then display it again. The second display always work.
  import ipywidgets.widgets
  out = widgets.Output()
  with out:
    display(solara_component())
  display(solara_component())

workarond_display(Page)
```

**Comment by dvn-signifyd** [comment]
Hi, I was using the workaround in the previous post successfully, recently the solutions found in this post started failing with the following error:
`Error displaying widget: Cannot read properties of undefined (reading 'extend')`
Same is happening for:
`box, rc = solara.render(Page())
box`
Can anyone else replicate?

**Reply by maartenbreddels** [reply]
Hi,

Was there anything that did work? For instance just rendering a normal widget? If not, you should contact databricks support for that.
I am happy to help, but they probably want to hear about this, and you are paying for their support on ipywidgets.

Regards,

Maarten

**Reply by dvn-signifyd** [reply]
Hi, thank you for replying!

I was checking all the options in the thread and it looks like the first one that doesn't work is 
```
import ipyvuetify as v
v.Btn(children=["Do you see me?"])
```
So I guess not all widgets are broken, but ipyvuetify is for some reason.

We've contacted DBX already, but hard to understand on which side the problem is at for me.
Is there anything I can run to help clarify what exactly is not working as intended in the DBX notebook environment?
Thanks in advance.

**Reply by maartenbreddels** [reply]
Another thing you could try is:
```
import ipyvue

ipyvue.Html(tag="div", children=["does this work?"])
```

This is the minimal thing that should work, so I would focus on getting this working. I wouldn't be surprised if third party widgets are broken at DBX. Another thing that could help, is downgrading ipyvue, in case we broke something in our stack

**Reply by banuatav** [reply]
@dvn-signifyd what Databricks runtime are you using? I have a similar issue as of this week. 

- Code that was working on on DBR 13.3 LTS ML last week returns error `Error displaying widget: Cannot read properties of undefined (reading 'extend')` now. 
- Judging by the small tests in the thread, ipyvuetify seems to be the culprit indeed. Downgrading ipyvue or ipyvuetify does not solve anything.
- The following does work:
```
import ipyvue

ipyvue.Html(tag="div", children=["does this work?"])
```

Going to report this issue to Databricks. Have you heard anything from them yet?

**Reply by dvn-signifyd** [reply]
Running DBR 14.3 and seeing the exact same failures you are, we have opened a DBX support ticket but no guarantees.

**Reply by tkrabel-db** [reply]
Hey @maartenbreddels - Tobi from Databricks here!

The support request worked! I am looking into this issue :) 
It seems that `ipyvuetify` (which used to work in Databricks) stopped working now too.

When creating an ipyvuetify widget in Databricks notebooks

```python
import ipyvuetify as v
my_select = v.Select(label='Fruits', items=['Apple', 'Pear', 'Cherry'])
my_select
```

I see the following error:
```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'extend')
    at ./src/mixins/themeable/index.ts (vuetify.js:37452:58)
    at n (vuetify.js:30:30)
    at ./src/components/VApp/VApp.ts (vuetify.js:391:75)
    at n (vuetify.js:30:30)
    at ./src/components/VApp/index.ts (vuetify.js:474:63)
    at n (vuetify.js:30:30)
    at ./src/components/index.ts (vuetify.js:30383:63)
    at n (vuetify.js:30:30)
    at ./src/index.ts (vuetify.js:32103:69)
    at n (vuetify.js:30:30)
```

I get the same error when running

```python
import ipyvuetify as v
v.Btn(children=["Do you see me?"])
```

Aside: the following code works:
```python
import ipyvue
ipyvue.Html(tag="div", children=["does this work?"])
```

I was curious what could be `undefined` and what should have the property `extend`. After some digging I found that vue 3 doesn’t have `Vue.extend` anymore ([src](https://github.com/vuejs/core/issues/629#issuecomment-575349384)). I think `ipyvuetify` uses a `vuetify` version that doesn’t work with vue ([stack overflow](https://stackoverflow.com/questions/60960059/vuetify-error-cannot-read-property-of-extend-of-undefined))

Does that ring a bell?

**Reply by maartenbreddels** [reply]
Hi 👋 

Can you confirm you are using ipyvue 1.x and ipyvuetify 1.x?
If so, you are still on Vue 2 and Vuetify 2.2.26
Looking at https://github.com/vuetifyjs/vuetify/blob/v2.2.26/packages/vuetify/src/mixins/themeable/index.ts#L5
I wouldn't be surprised that Vue itself is undefined. Maybe you can place a breakpoint somewhere and confirm that.

We make sure Vue gets loaded by this global widget: https://github.com/widgetti/ipyvue/blob/master/ipyvue/ForceLoad.py
Maybe a change at DBX does not load that widget anymore? 

For colab we saw a similar issue, and we use https://github.com/widgetti/reacton/blob/16518bf6fe9ca018bf8576398deb9c8bc27d6dc3/reacton/core.py#L2240 to get around that.

I hope that helps you track it down.

Regards,

Maarten

**Reply by jasongrout-db** [reply]
Tobi is looking into this more fully, but just to clarify @maartenbreddels - what do you mean by "Vue gets loaded by this global widget"? Are you loading the Vue library as a global in the browser namespace that you then expect to have access to from any other widget? If so, then I'm not surprised Databricks is running into the same symptoms you see in Colab, since both platforms use iframes to isolate outputs containing ipywidgets.

**Reply by jasongrout-db** [reply]
References about the colab issue and fix, to save others time looking them up:

* [Issue discussing the colab problem](https://github.com/googlecolab/colabtools/issues/3501)
* [Commit fixing the issue](https://github.com/widgetti/reacton/commit/8970afb21f51576fe15a63a333357535b9fc4e33#diff-bacbcded8a1b2c030b2888c001712c8d2765e6098cb601394337d22b5f3f7ee2L2144)

[From that colab issue](https://github.com/googlecolab/colabtools/issues/3501#issuecomment-1472328814):

> From blois "all display calls from the same cell are guaranteed to go into the same iFrame"

In Databricks, every top-level ipywidget renders in its own iframe (so there may be multiple iframes for a single cell output), so that is a difference between the two architectures that might complicate things.

**Reply by tkrabel-db** [reply]
@maartenbreddels thanks for the fast reply!

So I was checking the [issue reported on the colab github repo](https://github.com/googlecolab/colabtools/issues/3501) and [the fix](https://github.com/widgetti/reacton/commit/8970afb21f51576fe15a63a333357535b9fc4e33#diff-bacbcded8a1b2c030b2888c001712c8d2765e6098cb601394337d22b5f3f7ee2L2144) and I think this won't work for Databricks Notebooks since we render every widget in a separate iframe - there is no iframe consolidation logic yet.

I found a workaround with the following though:
```python 
import ipyvuetify as v
import ipyvue as vue
import ipywidgets
ipywidgets.VBox([
  vue.Html(tag="span", style_="display: none"), 
  v.Select(label='Fruits', items=['Apple', 'Pear', 'Cherry'])
])
```
knowing that this renders both widgets in the same iframe (@jasongrout-db need to correct myself; it does word).

I was wondering: can you wrap a component inside a box and add this Html span if we're on databricks notebooks? (I am happy to do it myself, but due to some migration, I cannot contribute to external repos atm)

You could add a case [here](https://github.com/widgetti/reacton/blob/8970afb21f51576fe15a63a333357535b9fc4e33/reacton/utils.py#L139-L153). You can identify Databricks Notebook through `get_ipython().__class__.__name__ == "DatabricksShell"`.

**Reply by maartenbreddels** [reply]
Hi Jason & Tobias,

> Are you loading the Vue library as a global in the browser namespace that you then expect to have access to from any other widget?

Yes. However, this worked before at Databricks. Maybe this helps:
https://github.com/widgetti/ipyvue/blob/9490c7b694414207cbb1ce8b15d993d80dda72bb/ipyvue/VueWidget.py#L143
Here we keep a reference to the global widget that all widgets have, which in principle should cause the widget manager to first load this widget. This is probably the reason it worked before, and might give a hint about what broke it.
